File: Python/121_BestTimeBuySellStock.py

class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        buy = prices[0]
        profit = 0

        for i in range(1, len(prices)):
            if prices[i] < buy:
                buy = prices[i]
            elif prices[i] - buy > profit:
                profit = prices[i] - buy 
        return profit    

test = Solution()
print(test.maxProfit([7,1,5,3,6,4])) # 5
print(test.maxProfit([7,6,4,3,1])) # 0

File: Java/src/_121BestTimeBuySellStock.java
public class _121BestTimeBuySellStock {
    public static int maxProfit(int[] prices) {
        int buy = prices[0];
        int profit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < buy) {
                buy = prices[i];
            }

            profit = Math.max(profit, prices[i] - buy);
        }
        return profit;
    }


    public static void main(String[] args) {
        int[] test = {7,1,5,3,6,4};
        int[] test1 = {7,6,4,3,1};
        System.out.println(maxProfit(test)); // 5
        System.out.println(maxProfit(test1)); // 0
    }
}
